src/dstack/_internal/server/utils/common.py

```diff
@@ -86,6 +86,23 @@ async def wait_to_lock(lock: asyncio.Lock, locked: Set[KeyT], key: KeyT, *, dela
         await asyncio.sleep(delay)
 
 
+async def wait_to_lock_many(
+    lock: asyncio.Lock, locked: Set[KeyT], keys: List[KeyT], *, delay: float = 0.1
+):
+    """
+    Retry locking until all the keys are locked.
+    Lock is released during the sleep.
+    """
+    left_to_lock = keys.copy()
+    while len(left_to_lock) > 0:
+        async with lock:
+            for key in left_to_lock:
+                if key not in locked:
+                    locked.add(key)
+                    left_to_lock.remove(key)
+        await asyncio.sleep(delay)
```

Comment:
I think a deadlock is possible. Please correct me if I'm wrong.

Task T1 wants to lock keys K1, K2.
Task T2 wants to lock keys K2, K1.
Task T3 holds the lock for key K2.

T1 locks K1 and waits until K2 is released.
T3 releases K2.
T2 locks K2 and waits until K1is released.

T1 and T2 are deadlocked.

Reply:
You're right. If we make it clear to the caller that keys should be sorted, then the deadlock should not be possible?

Reply:
I also think that we should always set timeout for wait_to_lock() and wait_to_lock_many() to prevent from deadlock but this should be done in a separate issue to carefully check for timeout consequences.

Reply:
> If we make it clear to the caller that keys should be sorted, then the deadlock should not be possible?

I think so. Or accept the keys as an unordered collection, e.g. set, and sort it inside `wait_to_lock_many`